src/tools/rbd/action/List.cc

```diff
@@ -56,139 +211,98 @@ int do_list(librbd::RBD &rbd, librados::IoCtx& io_ctx, bool lflag,
     tbl.define_column("LOCK", TextTable::LEFT, TextTable::LEFT);
   }
 
-  std::string pool, image, snap, parent;
+  for (int left = 1; left < std::min(threads, (int)names.size()); left++) {
+    worker_entry newworker;
+    r = init_worker(pool_name, newworker, rados);
+    workers.push_back(std::move(newworker));
+  }
 
-  for (std::vector<std::string>::const_iterator i = names.begin();
-       i != names.end(); ++i) {
-    librbd::image_info_t info;
-    librbd::Image im;
-
-    r = rbd.open_read_only(io_ctx, im, i->c_str(), NULL);
-    // image might disappear between rbd.list() and rbd.open(); ignore
-    // that, warn about other possible errors (EPERM, say, for opening
-    // an old-format image, because you need execute permission for the
-    // class method)
-    if (r < 0) {
-      if (r != -ENOENT) {
-        std::cerr << "rbd: error opening " << *i << ": " << cpp_strerror(r)
-                  << std::endl;
+  auto i = names.begin();
+  while (true) {
+    size_t workers_idle = 0; // if equal to workers.size(), then no more images to process
+    bool need_sleep = true; // true if we iterated over all workers and all are still processing
+    for (auto& comp : workers) {
+      switch (comp.opened) {
+	case LIST_IDLE:
+	  if (i == names.end()) {
+ 	    workers_idle++;
+ 	    continue;
+	  }
+	  comp.name = *i;
+	  comp.completion = new librbd::RBD::AioCompletion(nullptr, nullptr);
+	  r = comp.rbd->aio_open_read_only(*comp.io_ctx, *comp.img, i->c_str(), NULL, comp.completion);
+	  i++;
+	  comp.opened = LIST_OPENED;
+	  need_sleep = false;
+	  break;
+	case LIST_OPENED:
+	  if (comp.completion->is_complete()) {
+	    // image might disappear between rbd.list() and rbd.open(); ignore
+	    // that, warn about other possible errors (EPERM, say, for opening
+	    // an old-format image, because you need execute permission for the
+	    // class method)
+	    r = comp.completion->get_return_value();
+	    comp.completion->release();
+	    if (r < 0) {
+	      if (r != -ENOENT) {
+		std::cerr << "rbd: error opening " << *i << ": " << cpp_strerror(r)
+			  << std::endl;
+	      }
+	      // in any event, continue to next image
+	      comp.opened =  LIST_IDLE;
+	      continue;
+	    }
+	    r = list_process_image(&rados, comp, lflag, f, tbl);
```

Comment:
Yes, for sure. But I don't think it's a real problem, most users grep it or sort their way anyway, especially machine readable formats.